core/modules/util/ThreadPool.h

```diff
@@ -135,6 +158,26 @@ class ThreadPool : public std::enable_shared_from_this<ThreadPool> {
 
     EventThreadJoiner::Ptr _joinerThread; ///< Tracks and joins threads removed from the pool.
     std::atomic<bool> _shutdown{false}; ///< True after shutdownPool has been called.
+
+    // Functions to track the number of threads created by the pool.
+    /// Wait until the number of existing threads is <= max existing threads.
+    /// The pool kicks out threads that require significant time to process and
+    /// replaces them on a regular basis, see PoolEventThread::leavePool().
+    /// Under some circumstances, the number of threads that have been kicked out but
+    /// have not finished can become extremely large (tens of thousands). There
+    /// isn't much choice but to wait for some to finish before kicking more out.
+    void _waitIfAtMaxThreadPoolCount();
+
+    /// Increase the count of existing threads.
+    void _incrPoolThreadCount();
+
+    /// Decrease the count of existing threads.
+    void _decrPoolThreadCount();
+
+    unsigned int _poolThreadCount = 0;    ///< Number of threads that exist.
+    unsigned int _maxThreadCount = 20000; ///< Max number of thread allowed &&& use config file value
```

Comment:
This fixed number is not consistent with the default set to `10000` in the config file's parameter:
```
max_pool_threads = 10000
```